crates/sc-proof-of-time/src/state_manager.rs

```diff
@@ -0,0 +1,593 @@
+                .iter()
+                .find(|proof| proof.slot_number == received.slot_number)
+                .ok_or(PotConsensusStateError::ReceivedSlotMissing {
+                    summary: summary.clone(),
+                    slot_number: received.slot_number,
+                    block_number: format!("{block_number}"),
+                })?;
+            // Safe to index.
+            if *proof != *received {
+                return Err(PotConsensusStateError::ReceivedProofMismatch {
+                    summary: summary.clone(),
+                    slot_number: received.slot_number,
+                    block_number: format!("{block_number}"),
+                });
+            }
+
+            return Ok(());
+        }
+
+        // Check that the parent last proof and the block first proof
+        // form a chain.
+        let last_parent_proof =
+            parent_block_proofs
+                .iter()
+                .last()
+                .ok_or(PotConsensusStateError::ParentProofsEmpty {
+                    summary: summary.clone(),
+                    slot_number,
+                    block_number: format!("{block_number}"),
+                })?;
+        let first_block_proof =
+            block_proofs
+                .get(0)
+                .ok_or(PotConsensusStateError::ReceivedProofsEmpty {
+                    summary: summary.clone(),
+                    slot_number,
+                    block_number: format!("{block_number}"),
+                })?;
+        if first_block_proof.slot_number != (last_parent_proof.slot_number + 1) {
+            return Err(PotConsensusStateError::ReceivedUnexpectedSlotNumber {
+                summary: summary.clone(),
+                slot_number,
+                block_number: format!("{block_number}"),
+                expected: last_parent_proof.slot_number + 1,
+                actual: first_block_proof.slot_number,
+            });
+        }
+
+        // Compare the received proofs against the local chain. Since the
+        // local chain is already validated, not doing the AES check on the
+        // received proofs.
+        let mut expected_slot = first_block_proof.slot_number;
+        for received in block_proofs {
+            if received.slot_number != expected_slot {
+                return Err(PotConsensusStateError::ReceivedUnexpectedSlotNumber {
+                    summary: summary.clone(),
+                    slot_number,
+                    block_number: format!("{block_number}"),
+                    expected: expected_slot,
+                    actual: received.slot_number,
+                });
+            }
+            expected_slot += 1;
+
+            // TODO: avoid repeated lookups, locate start of range
+            let local_proof = state
+                .chain
+                .iter()
+                .find(|local_proof| local_proof.slot_number == received.slot_number)
+                .ok_or(PotConsensusStateError::ReceivedSlotMissing {
```

Comment:
Not sure I understand this. If we haven't seen a PoT for this slot we don't verify the proof received from the block? It seems like we should run AES verification here.

Reply:
Hm, we already verified the proof before adding to chain, if we have a proof for this slot in the chain. So we can just compare the fields, as the local chain is the source of truth.

May be I didn't understand the question

Reply:
I mean its possible that this block arrives to a node before the gossip. So the proofs are valid but we haven’t seen them yet

Reply:
> block arrives to a node before the gossip

I was looking into this last week as I was trying the end to end integration. Only way this can happen is the local PoT chain has some/none of the proofs from the consensus block being imported. All the proofs from the parent consensus block should already be in the chain (otherwise parent block won't have been verified).

For the case where we don't have all the PoT proofs yet, I am trying to figure out if 
1. We should fail import verification
2. Or if there is a way to say "don't fail, but don't have all the dependencies yet". 
For (1), don't want consensus to stall if PoT is lagging (may be momentarily). For (2), will try to find out if this is possible. If possible, how to trigger a retry when the PoT catches up

This will be part of the consensus integration change though

Reply:
> local PoT chain has some/none of the proofs from the consensus block being imported

Right. IMO when this happens we can "fill gaps" by verifying (AES) the block's PoT and setting the local PoT tip to the block's last proof. Because as soon as we receive a valid block with proofs up to slot `n`, the local chain up to slot `n` becomes useless to us as it is already in consensus chain. So we don't care to catch up via gossip.